File: web/src/lib/technologies.ts
import { api } from "@/lib/api";

// Technology usage across synced client accounts (account sync, 0121): per technology version,
// how many services / projects / clients / reachable people run it, their project spend, and the
// catalog lifecycle status (supported / deprecated / eol; unknown = not in the catalog).

export type TechStatus = "supported" | "deprecated" | "eol" | "unknown";

export interface TechnologyVersionUsage {
  technology: string;
  name: string;
  category: string;
  version: string;
  status: TechStatus;
  services: number;
  projects: number;
  companies: number;
  contacts: number;
  project_spend: number;
}

export async function fetchTechnologyUsage(): Promise<TechnologyVersionUsage[]> {
  return (await api<{ usage: TechnologyVersionUsage[] }>("/technologies")).usage;
}

/** Order versions numerically where possible (8.10 after 8.9), newest first. */
export function compareVersionsDesc(a: string, b: string): number {
  return b.localeCompare(a, undefined, { numeric: true });
}

/** Where the tenant's catalog comes from: "zerops" = Noola imports the public zerops.yml + import.yml
 *  schemas daily (unlisted in-use versions read as EOL); "manual" = pushed via the API; null = none. */
export interface CatalogState {
  source: "manual" | "zerops" | null;
  synced_at: string | null;
  last_error: string | null;
  technologies: number;
}

export async function fetchCatalogState(): Promise<CatalogState> {
  return (await api<{ catalog: CatalogState }>("/technologies/catalog")).catalog;
}

export async function setCatalogSource(source: "zerops" | "manual"): Promise<CatalogState> {
  return (await api<{ catalog: CatalogState }>("/technologies/catalog", { method: "PUT", body: JSON.stringify({ source }) })).catalog;
}

export async function syncCatalog(): Promise<CatalogState> {
  return (await api<{ catalog: CatalogState }>("/technologies/catalog/sync", { method: "POST" })).catalog;
}

File: sdk/go/technologies.go
package noola

import "context"

// TechnologyUsage is one technology version in use across your synced clients.
type TechnologyUsage struct {
	Technology string `json:"technology"`
	Name       string `json:"name"`
	Category   string `json:"category"`
	Version    string `json:"version"`
	// Status: "supported", "deprecated", "eol", "unknown" (not in a manual catalog). With the Zerops
	// catalog (Noola imports the zerops.yml + import.yml schemas itself) a version Zerops no longer
	// offers is "eol".
	Status       string  `json:"status"`
	Services     int     `json:"services"`
	Projects     int     `json:"projects"`
	Companies    int     `json:"companies"`
	Contacts     int     `json:"contacts"`
	ProjectSpend float64 `json:"project_spend"`
}

// TechnologyUsage returns usage per technology version.
func (c *Client) TechnologyUsage(ctx context.Context) ([]TechnologyUsage, error) {
	var out struct {
		Usage []TechnologyUsage `json:"usage"`
	}
	if err := c.do(ctx, "GET", "/v1/public/technologies", nil, nil, &out); err != nil {
		return nil, err
	}
	return out.Usage, nil
}

// TechnologyCatalog is a manually maintained catalog (PutTechnologies).
type TechnologyCatalog struct {
	Technologies []CatalogTechnology `json:"technologies"`
}

// CatalogTechnology is one catalog entry; Aliases fold spelling variants onto Key (golang → go).
type CatalogTechnology struct {
	Key      string           `json:"key"`
	Name     string           `json:"name,omitempty"`
	Category string           `json:"category,omitempty"`
	Aliases  []string         `json:"aliases"`
	Versions []CatalogVersion `json:"versions"`
}

// CatalogVersion is a version and its lifecycle status: "supported", "deprecated" or "eol".
type CatalogVersion struct {
	Version string `json:"version"`
	Status  string `json:"status"`
}

// PutTechnologies REPLACES the catalog with a manual one. You don't need this for Zerops: Noola
// imports the Zerops catalog itself (Technologies → "Use Zerops catalog"), and a manual push turns
// that import off.
func (c *Client) PutTechnologies(ctx context.Context, catalog TechnologyCatalog) error {
	for i := range catalog.Technologies {
		if catalog.Technologies[i].Aliases == nil {
			catalog.Technologies[i].Aliases = []string{}
		}
		if catalog.Technologies[i].Versions == nil {
			catalog.Technologies[i].Versions = []CatalogVersion{}
		}
	}
	return c.do(ctx, "PUT", "/v1/public/technologies", nil, catalog, nil)
}
